src/libexpr/print.cc

```diff
@@ -90,5 +119,286 @@ printAttributeName(std::ostream & str, std::string_view name) {
+            for (auto & i : sorted) {
+                if (attrsPrinted >= options.maxAttrs) {
+                    if (options.ansiColors)
+                        str << ANSI_FAINT;
+                    str << " «";
+                    pluralize(str, sorted.size() - attrsPrinted, "attribute", "attributes");
+                    str << " elided»";
+                    if (options.ansiColors)
+                        str << ANSI_NORMAL;
+                    break;
+                }
+
+                printAttributeName(str, i.first);
+                str << " = ";
+                if (seen && seen->count(i.second)) {
+                    if (options.ansiColors)
+                        str << ANSI_MAGENTA;
+                    str << "«repeated»";
+                    if (options.ansiColors)
+                        str << ANSI_MAGENTA;
+                } else
+                    print(str, *i.second, depth + 1, seen);
+                str << "; ";
+                attrsPrinted++;
+            }
+
+            str << "}";
+        } else
+            str << "{ ... }";
+
+        break;
+    }
+
+    case nList:
+        if (seen)
+            seen->insert(&v);
+
+        str << "[ ";
+        if (depth < options.maxDepth) {
+            size_t listItemsPrinted = 0;
+            for (auto elem : v.listItems()) {
+                if (listItemsPrinted >= options.maxListItems) {
+                    if (options.ansiColors)
+                        str << ANSI_FAINT;
+                    str << " «";
+                    pluralize(str, v.listSize() - listItemsPrinted, "item", "items");
+                    str << " elided»";
+                    if (options.ansiColors)
+                        str << ANSI_NORMAL;
+                    break;
+                }
+
+                if (seen && seen->count(elem)) {
+                    if (options.ansiColors)
+                        str << ANSI_MAGENTA;
+                    str << "«repeated»";
+                    if (options.ansiColors)
+                        str << ANSI_NORMAL;
+                } else {
+                    if (elem) {
+                        print(str, *elem, depth + 1, seen);
+                    } else {
+                        if (options.ansiColors)
+                            str << ANSI_MAGENTA;
+                        str << "«nullptr»";
+                        if (options.ansiColors)
+                            str << ANSI_NORMAL;
+                    }
+                }
+                str << " ";
+                listItemsPrinted++;
+            }
+        }
+        else
+            str << "... ";
+        str << "]";
+        break;
+
+    case nFunction:
+        if (v.isLambda()) {
+            if (options.ansiColors)
+                str << ANSI_BLUE;
+            if (v.lambda.fun) {
+                std::ostringstream s;
+                s << state.positions[v.lambda.fun->pos];
+                str << "«lambda @ " << filterANSIEscapes(s.str()) << "»";
```

Comment:
have you ever looked at the output of `nix-instantiate --eval --xml`? you're about 13 years too late with this request; and i'd consider `--xml` more stable than the "ATerm" output (which this still is according to the man page!).

Reply:
(Also, I think assuming that "coming up with the exact same position as C++nix" is both normal to want and possible to achieve is a weird assumption; especially in places where not even Nix specifies exactly what part of the attribute is pointed to by `unsafeGetAttrPos`. Or, well, unsafeGetAttrPos doesn't have any documentation at all, so...)

Reply:
Indeed, that is kind of unfortunate.

Reply:
> Also, I think assuming that "coming up with the exact same position as C++nix" is both normal to want and possible to achieve is a weird assumption

I don't think it's a weird aim to want to be able to produce the same `outPath`s as C++ Nix in an alternative implementations. Sadly, `unsafeGetAttrPos` for example leaks into derivations due to the manual IIRC. I'd say it's a valuable goal to keep the potential for such problems expanding small. `nix repl` is much less of an issue in this regard compared to `nix-instantiate`.

Reply:
In the case of `--xml`, it is possible to disable location printing via `--no-location`. Unfortunately, `--xml` (understandably) has no cycle detection, so testing that is not possible at all.

Reply:
> Unfortunately, `--xml` (understandably) has no cycle detection

Could be added behind an option if you/someone wants to use it for a test suite.

Reply:
I don't think having CLI options that are only really useful for the test suite is really wise. Cycle detection is not that important, since it is not technically a *language* test, rather a printing test. (In language, cycle detection would be impossible if pointer equality didn't exist…)